### PR TITLE
chore: Support range selection in Content Manager ListView Table.

### PR DESCRIPTION
**Pull request by gitstart**
Closes https://github.com/strapi/strapi/issues/15799

### What does it do?
Support range selection/deselection on ListView when holding the keyboard shortcut `Shift + Click`

### Why is it needed?
Prevents the need to manually range select/deselect one item at a time for a desired action.

### How to test it?

1. Go to `Content-Type Builder`
2. Create a test component with a text field name and save.
3. Go to `Content Manager` and add 3 to 4 entries to the `test` component.
4. Select the first item by checking the checkbox.
5. Skip the second item to the third and try holding down the Shift key while checking the third item.
6. Notice that range selection works.
7. With multiple items selected, try deselecting at a desired position while holding the Shift key, notice that it deselects the range.

### Related issue(s)/PR(s)

https://github.com/strapi/strapi/issues/15799

### Demo

https://user-images.githubusercontent.com/7220167/219575281-f2190ab9-1a20-40a2-8062-b78597eac365.mov

---
This code was written and reviewed by GitStart Community. Growing future engineers, one PR at a time.

### CODE REVIEW

**Review by joshuaellis — CHANGES_REQUESTED** [review]
This needs tests adding IMO to the `helper-plugin` component.

In addition you've changed the API so tests should also identify if the API additions would break if you only use the original parameters `name / value` - how can we make this backwards compat? 

Finally, you've only changed the table in the CM but I can see the DynamicTable component is used elsewhere in the app – it would make sense to mirror the functionality across the board :)

**Comment by gitstart** [comment]
Hi @joshuaellis,
Test has now been added to the `DynamicTable` component of `helper-plugin` kindly take a look.

File: packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  lightTheme,
+} from '@strapi/design-system';
+import { render, fireEvent, waitFor, screen } from '@testing-library/react';
+import * as React from 'react';
+import { IntlProvider } from 'react-intl';
+import { MemoryRouter } from 'react-router-dom';
+import TrackingProvider from '../../../providers/TrackingProvider';
+import Table from '../index';
+
+const TableRows = ({
```
Comment (joshuaellis) [inline packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:18]: How come you've chosen to re-write `TableRows`? this feels like we could change our code but this test would still pass?

**Review by joshuaellis — CHANGES_REQUESTED** [review]

File: packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js (hunk 1 above)
Comment (raph941) [reply packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:18]: Hi @joshuaellis,
Upon reviewing the layout of the codebase, we have observed that the import flow between the `helper-plugin` package and other packages that contain the `TableRow` component is unidirectional. We want to avoid introducing package circular import unless it's deemed the preferred approach. Noticing that the `TableRows` component was also re-written in the `DynamicTable.stories.mdx` file we opted to adopt the convention.

Furthermore, we have identified that there are currently two TableRow components that could potentially be imported into this test. Would appreciate your input on determining the optimal approach to resolve this situation.

Comment (gitstart) [reply packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:18]: Hi @joshuaellis,
Upon reviewing the layout of the codebase, we have observed that the import flow between the `helper-plugin` package and other packages that contain the `TableRow` component is unidirectional. We want to avoid introducing package circular import unless it's deemed the preferred approach. Noticing that the `TableRows` component was also re-written in the `DynamicTable.stories.mdx` file we opted to adopt the convention.

Furthermore, we have identified that there are currently two TableRow components that could potentially be imported into this test. Would appreciate your input on determining the optimal approach to resolve this situation.

Comment (joshuaellis) [reply packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:18]: Oh i see we don't export a TableRow. cool 👍🏼

File: packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+              rows={rows}
+              withMainAction={withMainAction}
+            >
+              <TableRows headers={headers} rows={rows} allowRangeSelection={allowRangeSelection} />
+            </Table>
+          </IntlProvider>
+        </ThemeProvider>
+      </TrackingProvider>
+    </MemoryRouter>
+  );
```
Comment (joshuaellis) [inline packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:85]: you should only render `Table/TableRows` the rest should use the `wrapper` option – https://testing-library.com/docs/react-testing-library/api/#wrapper

File: packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    { id: 2, firstname: 'm', lastname: 'frachet', email: 'm@strapi.io' },
+    { id: 3, firstname: 'hicham', lastname: 'ELBSI', email: 'helbsi@strapi.io' },
+    { id: 4, firstname: 'john', lastname: 'doe', email: 'doe@strapi.io' },
+    { id: 5, firstname: 'mary', lastname: 'jane', email: 'jane@strapi.io' },
+  ];
+
+  it('Performs single row item selection/deselection', async () => {
+    const { container } = renderTable(headers, rows);
+
+    expect(container).toMatchSnapshot();
```
Comment (joshuaellis) [inline packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:117]:
```suggestion
```
Please don't just snapshot it, instead assert that we have the headers as we expect etc.

File: packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    checkboxes.map((box) => expect(box).not.toHaveClass('selected'));
+
+    fireEvent.click(checkboxes[1]);
+    fireEvent.click(checkboxes[2]);
+
+    await waitFor(() => {
+      expect(checkboxes[1]).toHaveClass('selected');
+      expect(checkboxes[2]).toHaveClass('selected');
+    });
```
Comment (joshuaellis) [inline packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:129]: Can you use `userEvent` instead please? this should remove the `waitFor` calls.

File: packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+  ];
+
+  it('Performs single row item selection/deselection', async () => {
+    const { container } = renderTable(headers, rows);
+
+    expect(container).toMatchSnapshot();
+    const checkboxes = screen.getAllByRole('checkbox');
+    expect(checkboxes).toHaveLength(6);
+
+    checkboxes.map((box) => expect(box).not.toHaveClass('selected'));
```
Comment (joshuaellis) [inline packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:121]: This is an implementation detail, can you not check the `checked` attribute?

**Review by joshuaellis — CHANGES_REQUESTED** [review]
Some additional comments on your test, overall looking strong though 💪🏼 i've only written them once, but they apply to the whole test file.

File: packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js (hunk 3 above)
Comment (Boegie19) [reply packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:117]: @joshuaellis Just out of curiosity why is just snapshotting it not good enough.
Since it will check for an exact match and changing the snapshot would mean chancing the snapshot file aka
Aka if the snapshot has all the correct headers we expect etc. Then we are basically testing it 2 times.

Comment (joshuaellis) [reply packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:117]: In theory yes, they're good, but in practice i've (and others) have found they're not. No one reads / checks them and they're not notoriously hard to follow for HTML that's more than but a few lines especially as it brings in _all the classes_ and even then when you use styled-components classNames are like `c12` which will change with _most_ updates from the DS 🙃

If we wanted to implement visual regression testing i'd use something like chromatic because it's much more user friendly 😄

Comment (Boegie19) [reply packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:117]: Ye fair so it would be good for backend testing since there it does not change except if you tell it to but for frontend it is bad I see thanks for the explanation .

Comment (joshuaellis) [reply packages/core/helper-plugin/lib/src/components/DynamicTable/tests/Table.test.js:117]: > Ye fair so it would be good for backend testing since there it does not change except if you tell it to but for frontend it is bad I see thanks for the explanation .

yeah exactly, i think it's great for testing data-structures e.g. API responses!

**Review by joshuaellis — APPROVED** [review]
